src/net/http.rs

```diff
@@ -0,0 +1,2674 @@
+        Ok(to_send)
+    }
+
+    fn flush_chunk(&mut self) -> io::Result<usize> {
+        let sent = HttpChunkedTransferWriter::send_chunk(&mut self.fd, self.chunk_size, &self.chunk_buf)?;
+        self.chunk_buf.clear();
+        Ok(sent)
+    }
+
+    fn buf_chunk(&mut self, buf: &[u8]) -> usize {
+        let to_copy = 
+            if self.chunk_size - self.chunk_buf.len() < buf.len() {
+                self.chunk_size - self.chunk_buf.len()
+            }
+            else {
+                buf.len()
+            };
+
+        self.chunk_buf.extend_from_slice(&buf[0..to_copy]);
+        to_copy
+    }
+}
+
+impl<'a, W: Write> Write for HttpChunkedTransferWriter<'a, W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut written = 0;
+        while written < buf.len() {
+            if self.chunk_buf.len() > 0 {
+                if self.chunk_buf.len() < self.chunk_size {
+                    let nw = self.buf_chunk(&buf[written..]);
+                    written += nw;
+                }
+                if self.chunk_buf.len() >= self.chunk_size {
+                    self.flush_chunk()?;
+                }
+            }
+            else { 
+                if written + self.chunk_size < buf.len() {
+                    let nw = HttpChunkedTransferWriter::send_chunk(&mut self.fd, self.chunk_size, &buf[written..(written + self.chunk_size)])?;
+                    written += nw;
+                }
+                else {
+                    let nw = self.buf_chunk(&buf[written..]);
+                    written += nw;
+                }
+            }
+        }
+        Ok(written)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        // send out any bufferred chunk data
+        self.flush_chunk()
+            .and_then(|nw| {
+                if nw > 0 {
+                    // send empty chunk
+                    self.fd.write_all(format!("0\r\n\r\n").as_bytes())
+                        .and_then(|_nw| Ok(()))
+                }
+                else {
+                    Ok(())
+                }
+            })
+    }
+}
+
+impl<'a, W: Write> Drop for HttpChunkedTransferWriter<'a, W> {
+    fn drop(&mut self) -> () {
+        let _ = self.flush();
+    }
+}
+
+
+impl HttpRequestPreamble {
+    pub fn new(verb: String, path: String, hostname: String, port: u16, request_id: u32) -> HttpRequestPreamble {
+        HttpRequestPreamble {
+            verb: verb,
+            path: path,
+            host: PeerHost::from_host_port(hostname, port),
+            request_id: request_id,
+            content_type: None,
+            content_length: None,
+            headers: HashMap::new()
+        }
+    }
+
+    pub fn new_serialized<W: Write>(fd: &mut W, verb: &str, path: &str, host: &PeerHost, request_id: u32, content_length: Option<u32>, content_type: Option<&HttpContentType>, headers: &str) -> Result<(), net_error> {
+        // TODO: can we avoid allocating these?
```

Comment:
You should be able to just write to `fd`, rather than collecting them all to write at the end of the function, right?

Reply:
Yeah, will do that instead.

Reply:
Addressed in 3e15367d3